buildkit: add support for exporting build output with `--output` or `-o`

flouthoc opened the pull request:
Allows end-users to export final build content or rootfs to external formats.

By default, a local container image is created from the build result. The --output (or -o) flag allows you to override this behavior, and a specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation etc.

The value for --output is a CSV-formatted string defining the exporter type and options. Currently, local and tar exporters are supported. The local exporter writes the resulting build files to a directory on the client side. The tar exporter is similar but writes the files as a single tarball (.tar).

```console
buildah build --output type=local,dest=dir-rootfs .
buildah build --output type=tar,dest=rootfs.tar .
buildah build -o dir .
```
Reference: https://docs.docker.com/engine/reference/commandline/build/#custom-build-outputs

Closes: https://github.com/containers/buildah/issues/3789

TomSweeneyRedHat commented on docs/buildah-build.1.md:
```diff
@@ -407,6 +407,14 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+
+Output destination (format: type=local,dest=path)
+
+By default, a local container image is created from the build result. The --output (or -o) flag allows you to override this behavior, and a specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation etc.
```
```suggestion
By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation, etc.
```

flouthoc replied:
fixed

TomSweeneyRedHat commented on docs/buildah-build.1.md:
```diff
@@ -407,6 +407,14 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+
+By default, a local container image is created from the build result. The --output (or -o) flag allows you to override this behavior, and a specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation etc.
+
+The value for --output is a CSV-formatted string defining the exporter type and options. Currently, local and tar exporters are supported. The local exporter writes the resulting build files to a directory on the client side. The tar exporter is similar but writes the files as a single tarball (.tar).
```
```suggestion
The value for --output is a CSV-formatted string defining the exporter type and options. 

Valid _type_ values are:
- **local**: write the resulting build files to a directory on the client-side.
- **tar**: write the resulting files as a single tarball (.tar).
```

TomSweeneyRedHat replied:
An example or two would be helpful.  Looks like a good quick blog post once complete.

TomSweeneyRedHat replied:
do you handle a hyphen `-` with the tar option to write the output to stdout?  You might want to add that to the man page if so.

flouthoc replied:
Yes we handle `-` , added to docs with example.

TomSweeneyRedHat commented on tests/bud.bats:
```diff
@@ -545,6 +545,51 @@ _EOF
+  expect_output --substring 'invalid'
+  run_buildah 125 build --output type=wrong,dest=hello --signature-policy ${TESTSDIR}/policy.json -t test-bud -f $mytmpdir/Containerfile .
+  expect_output --substring 'invalid'
+}
```
Can you use the information from either file to build the image?  If so, could you add a test for that?

flouthoc replied:
Actually its equivalent to a `rootfs` on file-system so command does not supports this directly but user can always do something like, but its not supported by `buildah` and `buildkit`.
 ```Dockerfile
FROM scratch
ADD rootfs/ /
```
or Podman and Docker could directly use `import` instruction.

nalind commented on docs/buildah-build.1.md:
```diff
@@ -407,6 +407,19 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+
+Output destination (format: type=local,dest=path)
+
+By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a Container image, which can be useful for generating local binaries, code generation, etc.
```
No need to capitalize container here.

nalind commented on imagebuildah/executor.go:
```diff
@@ -352,6 +357,67 @@ func (b *Executor) resolveNameToImageRef(output string) (types.ImageReference, e
+	input, err := archive.Tar(mountPoint, archive.Uncompressed)
+	if err != nil {
+		return errors.Wrapf(err, "failed while performing tar for %q", mountPoint)
+	}
```
`input` is a `ReadCloser`, so it needs to be closed.

nalind commented on imagebuildah/executor.go:
```diff
@@ -352,6 +357,67 @@ func (b *Executor) resolveNameToImageRef(output string) (types.ImageReference, e
+			}
+			path = filepath.Join(cwd, filepath.Clean(path))
+		}
+		err = archive.Untar(input, path, nil)
```
How are ownerships intended to work for unprivileged users, where this happens inside of a user namespace?

flouthoc replied:
I'm expecting ownership to stay as it is which was set in image and it should not be changed for unprivileged users so image out can become read-only if files where meant to be written for some other user. 

But I'll verify behavior once with rootless buildkit as well just to stay in parity.

nalind replied:
This would need to be verified.

nalind replied:
What does that mean?  When we write content to a directory, who ends up owning files that are and aren't owned by root in the image?  Are setuid/setgid bits and extended attributes preserved?

flouthoc replied:
So when i export content with buildkit ownership does not stays as is everything is changing to current user so for example if I do 

```Dockerfile
FROM alpine
COPY hello .
RUN chown 1000:1000 hello
```

Exported content here has ownership of `hello` changed to `root:root` on host fs which I expected not to happen and stayed as `1000:1000` so just to match the behviour I added `NoLchown: true`.

flouthoc replied:
I have a hunch that either buildkit is not doing something right here and we are just trying to match the parity or maybe it is done intentionally to match some practical use-case.

flouthoc replied:
@nalind Following behavior also states true for `unprivileged user` an easier test is to run `buildkit:master-rootless` image.
So a reproducer is something like
```console
docker run --name buildkitd -it --privileged moby/buildkit:master-rootless
docker exec -it buildkitd sh
whoami
# should be `user` and not `root`
cd ~
touch hello
```

Inside container create containerfile

```Dockerfile
FROM alpine
COPY hello .
RUN chown 1000:1000 hello
```
and run
```
buildctl build --frontend=dockerfile.v0  --local context=. --local dockerfile=. --output type=local,dest=output-dir`
```
all the files inside `output-dir` have been chowned to user `user:user`.


Currently buildah follows this exact same behavior.

nalind replied:
Are setuid/setgid bits and extended attributes preserved?

flouthoc replied:
@nalind Yes both are preserved in both buildkit and buildah verified manually with `setfattr` and `getfattr` after unpacking.

flouthoc replied:
Just adding a note here with a reproducer shared by @nalind , `buildkit` has inconsistent and not documented behavior between `privileged` and `unprivileged` invocation and permissions are produced differently for binaries like `/usr/bin/write` in both invocation.

So something like 
```Dockerfile
FROM registry.fedoraproject.org/fedora:35
RUN rpm --setcaps shadow-utils
RUN dnf -y install attr util-linux && dnf clean all
RUN chmod u+rwx `rpm -ql filesystem` 2> /dev/null || true
RUN echo test > /test.txt && chown 1000:1000 /test.txt
RUN getcap -v /usr/bin/newuidmap /usr/bin/newgidmap ; stat -c "%A   %n" /test.txt /usr/bin/su /usr/bin/write /usr/bin/sudo
``` 

With buildkit exported output
```
#root
-rwxr-sr-x usr/bin/write
# unprivileged user
-rwxr-xr-x usr/bin/write
```

With buildah we share permissions similarly between invocation by privileged and unprivileged user 
```
#root
-rwxr-sr-x   usr/bin/write
# unpriviledged user
-rwxr-sr-x   usr/bin/write
```

Here behavior of buildah is different from buildkit a little bit but I was unable to find a reason why should we alter permission for unprivileged user hence leaving a note here so in future if we ever face this we can revisit this discussion.

flouthoc replied:
Still not entirely sure about what should we do here and which behavior is the correct one.

flouthoc replied:
Looked at it again it seems like the issue is with `unpriviledged` invocation of buildkit it does not preserves `setuid` but preserves `setgid` when manually added in Dockerfile but again overrides `setgid` for the files which were already there in the base image. This behavior looks really confusing to me could it be a bug in buildkit ? 

`priviledged` invocation of both buildah and buildkit works in the same manner and `unpriviledged` invocation of buildah preserves everything.

@nalind can we hit any security issues if we choose not to replicate buildkit's unpriviledged invocation behavior ?

nalind replied:
Either there's some rule there that we don't understand, or it's a bug.  `sudo` seems to be an outlier, as I also lose the setuid bit on binaries like passwd (from the passwd package), chfn and chsh (from the util-linux-user package), and the setgid bit on binaries like locate (from the mlocate package) and lockdev (from the lockdev package) when I run a build as an unprivileged user.
The general case sure seems to suggest that we should be stripping the setuid/setgid/sticky bits when we're writing to "local" destinations as an unprivileged user.

nalind replied:
Also, pretty sure we want to be preserving ownership information when we're invoked as root, even though buildkit seems to change setgid-"tty" /usr/bin/write and setgid-"lock" /usr/sbin/lockdev into setgid-"root" /usr/bin/write and /usr/sbin/lockdev.

flouthoc replied:
@nalind Modified buildah to preserve ownership as is when invoked with as `root`, this is different behavior from buildkit but looks good to me.

flouthoc replied:
They are preserved as is.

nalind replied:
The buildkit version also seems to create the top-level destination directory with 0700 permissions if it doesn't already exist, while we seem to create it as 0755 (neither modifies the permissions on an already-present directory).  If we're going to be putting setuid/setgid files in there, I'd be more comfortable if we used 0700.

flouthoc replied:
nice catch, fixed that as well.

nalind replied:
I still think that what we're seeing with `sudo`'s setuid bit being preserved when buildkit is run by an unprivileged user is not what we want to be using as our rule for everything.  We should be stripping setuid/setgid bits and capability xattrs when we're writing to "local" destinations as a non-root user.

flouthoc replied:
Pushed a new version which covers this for `non-root` user.

flouthoc commented:
> This is the sort of thing that we've historically done in the image library and then exposed via the `-t` flag, as copying images is a big part of what the image library does, and I could definitely see a `-o` implementation leaning on that as well. There's a "tarball" `ImageTransport`, that currently only implements an `ImageSource` that imports rootfs tarballs, that this would specifically appear to complement. The current PR isn't implemented that way, and that limits our ability to reuse this functionality anywhere else.

@nalind I agree with this, actually i thought of it but i was reluctant of moving this function to `c/image` cause I was not able to think of lot of use-cases outside of the build output generation. Since it looked like a buildah local to me hence i opted to keep it here but we can move this function to `c/image` in future if more cases show up or now as well whatever sounds best.

TomSweeneyRedHat commented:
If you think there's a chance you'd want to move some of this code in the future to c/image, I'd do it now.  Far less pain.  Regardless, I'd move it based on Nalin's comments.

rhatdan commented:
@vrothberg @mtrmac Thoughts?

mtrmac commented on docs/buildah-build.1.md:
```diff
@@ -407,6 +407,19 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+
+By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a container image, which can be useful for generating local binaries, code generation, etc.
+
+The value for --output is a CSV-formatted string defining the exporter type and options.
```
CSV is a specific spreadsheet format. This is not quite that.

The dest= option is only documented with an example… maybe go all the way and say that it is a comma-separated sequence of key=value pairs, and make a list/table of the supported keys.

flouthoc replied:
fixed.

mtrmac commented on docs/buildah-build.1.md:
```diff
@@ -820,6 +833,16 @@ buildah bud --platform linux/arm64 --platform linux/amd64 --manifest myimage /tm
 
+### Building an image using (--output) custom build output
+
+buildah build -o out .
```
(Personal rant: Ugh, so the syntax of --output is a set of key-value pairs, _except when it’s not_? I hate ambiguities.)

flouthoc replied:
this is just to match the short-hand syntax which `docker buildx` supports.

nalind replied:
The integration test should probably check that we don't unintentionally start rejecting this variation.

mtrmac replied:
(Marked unresolved to add that test, or to explicitly decide that we don’t need one.)

flouthoc replied:
added a test.

mtrmac commented on imagebuildah/executor.go:
```diff
@@ -352,6 +357,68 @@ func (b *Executor) resolveNameToImageRef(output string) (types.ImageReference, e
+	}
+	defer input.Close()
+	if isDir {
+		if !filepath.IsAbs(path) {
+			cwd, err := os.Getwd()
+			if err != nil {
+				return err
+			}
+			path = filepath.Join(cwd, filepath.Clean(path))
+		}
```
Wouldn’t it make sense to do this sometime early, during option processing or setup, so that we don’t do all the build work only for the saving to fail on a trivial path error?

flouthoc replied:
yes makes sense to me, fixed in latest commit.

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,56 @@ var (
+	}
+	if strings.Contains(buildOutput, ",") {
+		tokens := strings.SplitN(buildOutput, ",", 2)
+		if len(tokens) != 2 {
```
Is this somehow a fundamental design constraint? A compatibility constraint? The format at seems to support an arbitrary number of options just fine.

If this is to support a `,` character inside a path, please explicitly document that in a comment here (and perhaps even in the man page). (But I’m not sure about that idea, it limits us from adding more option types in the future.)

flouthoc replied:
We are now erroring out for every unknown option

mtrmac replied:
That’s not quite what I meant.

With this, `type=tar,dest=/tmp/with,option=comma` parses the path as `/tmp/with,option=comma`, which is both undocumented and making it impossible for us to add any more options (after a `dest` option, or ever? undocumented behavior).

I’d suggest using SplitN(…3) at least, to forbid commas in values, or preferably just using `Split()`, removing this length check completely, and relying on the duplicate checks, and on after-the-loop checks to verify all data has been provided.

flouthoc replied:
fixed.

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,56 @@ var (
+				}
+			}
+			if arr[0] == "dest" && !pathSelected {
+				pathSelected = true
```
It would be helpful for this to loudly complain about duplicate type/dest keys, instead of silently ignoring data.

flouthoc replied:
yes added a check for duplicates.

mtrmac commented on imagebuildah/executor.go:
```diff
@@ -776,6 +843,17 @@ func (b *Executor) Build(ctx context.Context, stages imagebuilder.Stages) (image
 	logrus.Debugf("printing final image id %q", imageID)
+	if b.buildOutput != "" {
+		logrus.Debugf("Generating custom build output with options %q", b.buildOutput)
+		path, isDir, isStdout, err := internalParse.GetBuildOutput(b.buildOutput)
```
Wouldn’t it be better to parse all of this (and report errors) before starting to actually build anything?

(And then, the result of `GetBuildOutput` should probably be an object that can just be passed on as a single unit, not a set of loose independent variables the caller needs to manage one at a time.)

flouthoc replied:
This is a nice idea, added in latest commit.

mtrmac commented:
> This is the sort of thing that we've historically done in the image library and then exposed via the `-t` flag, as copying images is a big part of what the image library does, and I could definitely see a `-o` implementation leaning on that as well.
> There's a "tarball" `ImageTransport`, that currently only implements an `ImageSource` that imports rootfs tarballs, that this would specifically appear to complement. The current PR isn't implemented that way, and that limits our ability to reuse this functionality anywhere else.

I definitely see an elegance in turning this option into just a “commit” to a special transport — because various options could be reused, and the Buildah code could be simpler.

OTOH there are also downsides:
- It would also allow the user to specify the other commit options/flags/features that modify the config, while the config is not produced by this tarball output. It would be extra code to reject all of that (… admittedly this PR should _also_ be rejecting all of those options with `--output` (unless `-t` is used?)).
- The c/image references are a bit stateless, making it somewhat ugly to create a reference “to stdout”. Buildah is already dealing with that for the tarball input format, OTOH, so that would not be a new concern.
- Note that c/image really has no concept of layer _contents_; and no facility for layer editing. I.e. to output a single tarball, it would have to be Buildah (via it’s `CommitOptions.Squash` code that would turn a multi-layer image into a tarball content.  (Alternatively, the tarball transport’s destination could reimplement all of that, by extracting the tarballs, implementing whiteouts, and then creating a tarball out of the result, but that would be a lot of unwarranted complexity at probably much worse performance than `extractRootfs` doing a `Mount`). Essentially, all the c/image tarball transport would do is to stream the Buildah-created tarball from a Buildah-created temporary file (one that is necessary to implement `containerImageRef.NewImageSource`) to the specified destination, hopefully without introducing a _second_ temporary file; while the current special-cased implementation can stream data from `archive.Tar` directly. So using a tarball transport would be an extra cost.

---

Overall I don’t really have a strong opinion about using a c/image transport.

I do, however, feel somewhat strongly (unjustifiably strongly for a Buildah outsider, perhaps) that the code to produce these tarballs should be shared with the code that produces tarballs in squashed images (at least including `containerImageRef.extractRootfs` — see how it differs from this PR’s implementation — possibly more). At least the “producer” half of `buildCustomOutput` should not exist IMHO.

Turning the `--output` option into just a transport reference would clearly be a way to do that. Doing it some other way, to preserve the streaming implementation without temporary files, might well be worthwhile — it’s a code size / long-term maintenance effort vs. performance tradeoff.

flouthoc commented:
@mtrmac I am fine to share as much code as possible with https://github.com/containers/buildah/blob/main/image.go#L153 `extractRootfs`. @nalind WDYT ?

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,70 @@ var (
+					return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid type %q selected for build output options %q", arr[1], buildOutput)
+				}
+			} else if arr[0] == "type" && pathSelected {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("Duplicate %q not supported", arr[0])
```
`"type" && pathSelected` seems incorrect.

flouthoc replied:
fixed

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,70 @@ var (
+			} else if arr[0] == "dest" && pathSelected {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("Duplicate %q not supported", arr[0])
+			}
+			if !typeSelected && !pathSelected {
```
This works but it’s not quite obvious that it is the “unknown option name” requires maintaining this list of *Selected variables.

Also the two separate checks for `arr[0] == $value` seem unnecessary.

How about
```go
switch arr[0] {
case "type":
    if typeSelected { fail }
    typeSelected = true
    …
case "dest":
    if pathSelected { fail }
    pathSelected = true
    …
default:
    fail
}
```

or even
```go
// fieldsSeen := set<string> , I’m too lazy to use actual Go here. I guess a map[string]struct{}.
field := arr[0]
if field in fieldsSeen { fail }
fieldsSeen.Add(field)
switch field {
case "type":
    …
case "dest":
    …
default:
    fail
}
```

flouthoc replied:
fixed

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,70 @@ var (
+		for _, option := range tokens {
+			arr := strings.SplitN(option, "=", 2)
+			if len(arr) != 2 {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid build output options %q", buildOutput)
```
I wouldn’t bother with filling in the `BuildOutputOption` values on error paths — the values are not expected to be used, and it only makes the `return`s look like success cases at a first glance.

flouthoc replied:
fixed

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,70 @@ var (
+		}
+	} else {
+		// expect default --output <dirname>
+		return BuildOutputOption{buildOutput, isDir, isStdout}, nil
```
Please move this short to be the first case in the if, much closer to the condition.  It can even be an early return and then the comma-separated syntax does not need to be indented — both the `if` above _and_ the code below only apply to the comma-separated syntax, and it is unclear why they should be indented differently and split into two parts with this intermediate code.

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,70 @@ var (
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("Duplicate %q not supported", arr[0])
+			}
+			if !typeSelected && !pathSelected {
+				return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid build output options %q", buildOutput)
```
Please tell the user about what is wrong about the input.

mtrmac commented on internal/parse/parse.go:
```diff
@@ -37,6 +37,70 @@ var (
+	}
+
+	if !typeSelected || !pathSelected {
+		return BuildOutputOption{path, isDir, isStdout}, fmt.Errorf("invalid build output options %q", buildOutput)
```
Please tell the user about what is wrong about the input.

flouthoc replied:
fixed.

mtrmac commented on cmd/buildah/build.go:
```diff
@@ -371,6 +378,7 @@ func buildCmd(c *cobra.Command, inputArgs []string, iopts buildOptions) error {
 		OS:                      systemContext.OSChoice,
 		Out:                     stdout,
 		Output:                  output,
+		BuildOutput:             iopts.BuildOutput,
```
Doing the parsing in `GetBuildOutput` above is nice; actually providing the result of the parsing here, instead of forcing the caller to parse again, would be cleaner. It avoids unreachable error cases in the ultimate implementation, and makes it much less likely to confuse the string syntax and the parsed data.

mtrmac replied:
I do think that all such tools should do their own parsing; that way they will fail quickly on invalid input, rather than triggering the long and costly build, and only failing in the commit step. So if they do the parsing, they might just as well provide the parsed value, and that eliminates a failure case in the commit step.

Also consider the podman-remote case, where the remote server basically ignores the users’ input and needs to provide a `BuildOutput` value pointing at a temporary tarball location/stream. If the options allow parsing a structure, the server can just pass a `BuildOutputOption{Path:theTemporaryFile, …}`. If the options require a string, the server will need to, quite pointlessly, format one.

This is not blocking from my perspective, and there might be something that makes passing the deconstructed value unattractive. Still, I have a preference :)

mtrmac replied:
> I do think that all such tools should do their own parsing

That implies that `GetBuildOutput` would have to be in an externally-importable package, not in `buildah/internal`.

flouthoc replied:
moved this to external parse pkg so it can be used by external parties if needed.

rhatdan commented:
@flouthoc Any progress on this?

flouthoc commented:
I'll fix the requested nits, thanks @mtrmac

rhatdan commented:
@flouthoc What is going on with this one?

flouthoc commented:
Picking this up again, i don't know how i missed this. :\

flouthoc commented on imagebuildah/executor.go:
```diff
@@ -776,6 +843,17 @@ func (b *Executor) Build(ctx context.Context, stages imagebuilder.Stages) (image
 		return "", nil, err
 	}
 	logrus.Debugf("printing final image id %q", imageID)
+	if b.buildOutput != "" {
```
Assuming I understood this comment correctly, docker allows `-t` and `-o` to be functional at same time independently and `--output` is functional even if `-t` is not specified so just keeping in parity with that.

Please unresolve this part if I understood this comment wrong.

flouthoc commented:
@mtrmac @nalind Addressed all feedbacks and also switched to a new approach after discussion above https://github.com/containers/buildah/pull/3823#issuecomment-1073546133 we are now reusing `containerImageRef.extractRootfs` https://github.com/containers/buildah/blob/main/image.go#L153 and no new implementation is added for extracting `Rootfs`

PTAL

rhatdan commented:
@mtrmac @nalind PTAL
Would like to get this in before the beta.

nalind commented on cmd/buildah/build.go:
```diff
@@ -328,6 +329,12 @@ func buildCmd(c *cobra.Command, inputArgs []string, iopts buildOptions) error {
 		timestamp = &t
 	}
+	if c.Flag("output").Changed {
+		buildOption, _ := internalParse.GetBuildOutput(iopts.BuildOutput)
```
This needs to handle errors.

nalind commented on docs/buildah-build.1.md:
```diff
@@ -411,6 +411,24 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+
+Output destination (format: type=local,dest=path)
+
+By default, a local container image is created from the build result. The --output (or -o) option allows you to override this behavior and specify a custom exporter. For example, custom exporters allow you to export the build artifacts as files on the local filesystem instead of a container image, which can be useful for generating local binaries, code generation, etc.
```
This suggests that when this option is used, we don't commit a container image, but it looks like we still do.

flouthoc replied:
Fixed explicitly mentioned.

nalind replied:
This still reads to me as though using this option causes the builder to not commit an image, as committing the image is the default behavior that's being overridden when this option is specified.

flouthoc replied:
fixed it after confirming from @nalind hence marking resolved again.

nalind commented on image.go:
```diff
@@ -813,3 +813,7 @@ func (b *Builder) makeImageRef(options CommitOptions) (types.ImageReference, err
 	return ref, nil
 }
+
+func (b *Builder) MakeImageRef(options CommitOptions) (types.ImageReference, error) {
```
This method doesn't need to exist.  Just have `Commit()` call `MakeContainerImageRef()` directly.

flouthoc replied:
`Commit(` is expecting type `types.ImageReference` but `MakeContainerImageRef` returns `*ContainerImageRef`, function is just helping me avoid recasting in the code but I'm can change this if there is better casting method.

flouthoc replied:
this is fixed.

nalind commented on internal/parse/parse.go:
```diff
@@ -41,6 +41,67 @@ var (
+	}
+	if !strings.Contains(buildOutput, ",") {
+		// expect default --output <dirname>
+		return BuildOutputOption{buildOutput, isDir, isStdout}, nil
```
Might as well name the fields in the initializers and use constants for the values, as above.

nalind commented on internal/parse/parse.go:
```diff
@@ -41,6 +41,67 @@ var (
+	for _, option := range tokens {
+		arr := strings.SplitN(option, "=", 2)
+		if len(arr) != 2 {
+			return BuildOutputOption{}, fmt.Errorf("invalid build output options %q", buildOutput)
```
We know exactly why we're rejecting the provided value here, so the error message doesn't need to be that vague.

nalind commented on internal/parse/parse.go:
```diff
@@ -41,6 +41,67 @@ var (
+			pathSelected = true
+			path = arr[1]
+		default:
+			return BuildOutputOption{}, fmt.Errorf("invalid build output options %q", buildOutput)
```
Likewise, we know `arr[0]` is not a recognized key for this option, as below.

nalind commented on internal/util/util.go:
```diff
@@ -22,3 +28,35 @@ func LookupImage(ctx *types.SystemContext, store storage.Store, image string) (*
+		path = filepath.Join(cwd, filepath.Clean(path))
+	}
+	if isDir {
+		err = archive.Untar(input, path, nil)
```
Use `chrootarchive.Untar()` here if at all possible.

flouthoc commented:
Fixed all the nits @rhatdan @nalind

nalind commented:
How will `podman-remote` be able to wire this up?

flouthoc commented:
@nalind I think it can be used by `podman-remote` only when output is exported on shared-mount path with the host.

nalind commented:
> @nalind I think it can be used by `podman-remote` only when output is exported on shared-mount path with the host. Initially this can be `NO-OP` for remote but can be enabled when `host`->`machine` sharing starts working on `macOS`.

Just so long as this doesn't lead to letting people who are accessing a server that only offers them the podman API using this to overwrite configuration files on the host.

rhatdan commented:
Probably should be rejected for now.

flouthoc commented:
> Just so long as this doesn't lead to letting people who are accessing a server that only offers them the podman API using this to overwrite configuration files on the host.

Yeah this might be not possible since there is no way to transfer files from server to host so its a `no-op` for `podman-remote` as of now so this scenario is not possible

> Probably should be rejected for now.

@rhatdan Do you mean putting this as `no-op` for `podman-remote` or getting this feature into `podman` only when it works for remote ?

flouthoc commented:
@rhatdan @nalind resolved all nits PTAL. Just `podman-remote` discussion is remaining but I think it can be `no-op` for now and we can discuss it podman side.

rhatdan commented:
Yes for podman-remote, I think we should reject it until we have a solid solution for it.  The problem with that is on a MAC and Windows.  We can discuss this at Water Cooler, to make us all understand the issue.

rhatdan commented:
LGTM

nalind commented on imagebuildah/stage_executor.go:
```diff
@@ -1556,6 +1568,23 @@ func (s *StageExecutor) commit(ctx context.Context, createdBy string, emptyLayer
+	// generate build output
+	if s.executor.buildOutput != "" {
+		buildOutputOptions := options
+		src, err := s.builder.MakeContainerImageRef(buildOutputOptions)
```
What's the temporary variable for?

flouthoc commented:
We found one discrepancy in buildkit's behavior and its discussed and documented here: https://github.com/containers/buildah/pull/3823#discussion_r853821897, as of now I am not sure why that discrepancy exists and i'm unable to think of a reason to have that in buildah. I think we can keep that comment open and I'll address that in future in a followup PR if that becomes a problem.

mtrmac commented on internal/util/util.go:
```diff
@@ -22,3 +29,35 @@ func LookupImage(ctx *types.SystemContext, store storage.Store, image string) (*
+		if err != nil {
+			return err
+		}
+		path = filepath.Join(cwd, filepath.Clean(path))
```
`filepath.Join` implicitly runs `Clean` on the _output_. Does calling `Clean` on the _input_ make any non-obvious difference? If so, please document that.

mtrmac commented on internal/util/util.go:
```diff
@@ -22,3 +29,35 @@ func LookupImage(ctx *types.SystemContext, store storage.Store, image string) (*
 }
+
+// ExportFromReader reads bytes from given reader and exports to external tar, directory or stdout.
+func ExportFromReader(input io.Reader, isDir bool, isStdout bool, path string) error {
```
Couldn’t this just accept a `BuildOutputOption` argument, simplifying the only caller, and completely eliminate the risk of mismatching the order of the bool options?

mtrmac commented on tests/bud.bats:
```diff
@@ -551,6 +551,51 @@ _EOF
+  run_buildah build --output type=tar,dest=$mytmpdir/rootfs.tar --signature-policy ${TESTSDIR}/policy.json -t test-bud -f $mytmpdir/Containerfile .
+  # explode tar
+  mkdir $mytmpdir/rootfs
+  tar -C $mytmpdir/rootfs -xvf $mytmpdir/rootfs.tar
```
(If the goal is to only check file presence, using `tar -t` would allow listing the contents of the tarball without writing to disk. OTOH this actually verifies that the archive can be correctly extracted (assuming the tests fail on any failed command), which is valuable.)

flouthoc replied:
I didn't thought of any advantage here but I like the explanation on the second part hence keeping it as it is please feel free to unresolve this comment.

mtrmac commented on imagebuildah/stage_executor.go:
```diff
@@ -1556,6 +1568,22 @@ func (s *StageExecutor) commit(ctx context.Context, createdBy string, emptyLayer
+		if err != nil {
+			return "", nil, errors.Wrapf(err, "error computing layer digests and building metadata for container %q", s.builder.ContainerID)
+		}
+		rc, _, err := src.ExtractRootfs()
```
Why is it OK to ignore `errChan`?

(Alternatively, I could see an argument that `errChan` should not exist, and instead `rc.Close()` should be blocking not that channel and handling the error, and that users of `rc` should always be obliged to check the return value of `rc.Close()`. That feels conceptually clearer but might just be a lot of extra complexity for no benefit.)

flouthoc replied:
handled.

mtrmac commented on internal/parse/parse.go:
```diff
@@ -41,6 +41,71 @@ var (
+		return BuildOutputOption{}, fmt.Errorf("invalid build output options %q, accepted keys are type and dest", buildOutput)
+	}
+
+	return BuildOutputOption{path, isDir, isStdout}, nil
```
Name the fields in this initializer as well, please.

mtrmac reviewed (commented):
Just a few non-trivial things, and a lot of tiny text/documentatino nits… But also see elsewhere.

mtrmac commented:
> Yes for podman-remote, I think we should reject it until we have a solid solution for it. The problem with that is on a MAC and Windows. We can discuss this at Water Cooler, to make us all understand the issue.

For the record, the water caller discussion pointed out that the remote load/save implementations stream a tarball, e.g. https://github.com/containers/podman/blob/712c3bb226a84c8a5593d62648975611cea7839d/pkg/domain/infra/tunnel/images.go#L259 . OTOH build is somewhat special, as an outsider I’m not sure that is viable here.

IIRC we didn’t discuss the intermediate situation, but just not implementing this on remote in the first version (or do we have to explicitly filter that option out in the server?) seems easiest.

mtrmac commented:
> We found one discrepancy in buildkit's behavior and its discussed and documented here: [#3823 (comment)](https://github.com/containers/buildah/pull/3823#discussion_r853821897), as of now I am not sure why that discrepancy exists and i'm unable to think of a reason to have that in buildah. I think we can keep that comment open and I'll address that in future in a followup PR if that becomes a problem.

Set-UID executables are Serious Stuff. I’m _very uncomfortable_ with shipping that feature without making a detailed decision, as well as with the idea of shipping one behavior (and having users possibly test that and rely on that), and possibly later changing that.

(I have no idea what the right behavior is here, neither from the compatibility perspective nor from the security perspective. I just really dislike to the idea of “this is hard, let’s declare success anyway” in this aspect.

As a possible point of comparison, what does `podman copy` do when copying out other-UID/set-UID/otherwise unusual files out of a container? Shouldn’t that be consistent with this code?)

flouthoc commented on docs/buildah-build.1.md:
```diff
@@ -824,6 +842,16 @@ buildah bud --platform linux/arm64 --platform linux/amd64 --manifest myimage /tm
 
+### Building an image using (--output) custom build output
+
+buildah build -o out .
```
Edited man page above

flouthoc commented on internal/parse/parse.go:
```diff
@@ -41,6 +41,71 @@ var (
+	typeSelected := false
+	pathSelected := false
+	path := ""
+	if len(buildOutput) == 1 && buildOutput == "-" {
```
Handles `dest=-` now.

flouthoc commented on internal/parse/parse.go:
```diff
@@ -41,6 +41,71 @@ var (
+		// expect default --output <dirname>
+		return BuildOutputOption{Path: buildOutput,
+			IsDir:    isDir,
+			IsStdout: isStdout}, nil
```
fixed

flouthoc commented on image.go:
```diff
@@ -43,7 +43,7 @@ const (
 )
 
-type containerImageRef struct {
+type ContainerImageRef struct {
```
We can move `ContainerImageRef` to internal but as of now I don't see lot of harm in making it public since it has access to some useful methods and ContainerImageRef does not changes frequently but making it internal is safer.

nalind replied:
(In the case of this one in particular, the last revision of the PR I saw needed to be able to call `ExtractRootFS` from a different package, and I didn't have a better way to accomplish that in-hand.)

nalind replied:
Okay, an alternative would be to leave the `containerImageRef` type and its `extractRootfs()` method private, and add a public method to `Builder` that internally calls `makeContainerImageRef()` to create a `containerImageRef`, calls the resulting ref's `extractRootfs()` method, and hands the result back to the new methods's caller.  That's enough for the use case in `StageExecutor.commit()`, and doesn't expose a lot of our internals, so I think I like it better.

flouthoc replied:
ackd SGTM I'll add these changes.

flouthoc replied:
fixed.

flouthoc commented:
@mtrmac Thanks for review resolved all the nits. For this ( https://github.com/containers/buildah/pull/3823#issuecomment-1104239909 ) i agree however i really don't understand the disparity with the buildkits output for `root` and `rootless` invocation if we decide to differ from buildkit here we can close down on one consistent behavior. I'll discuss this with @nalind whenever he is back and do the needful.

mtrmac commented on docs/buildah-build.1.md:
```diff
@@ -411,6 +411,24 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+- **local**: write the resulting build files to a directory on the client-side.
+- **tar**: write the resulting files as a single tarball (.tar).
+
+If no type is specified, the value defaults to local. Use a hyphen as non-comma format (`-`) to write the output tarball to standard output (`STDOUT`).
```
Maybe
```suggestion
If no type is specified, the value defaults to local. Use a hyphen as non-comma format (`-`) to write the output tarball to standard output.
```
There is no user-facing `STDOUT` identifier / name, to my knowledge.

mtrmac commented on docs/buildah-build.1.md:
```diff
@@ -411,6 +411,24 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+- **local**: write the resulting build files to a directory on the client-side.
+- **tar**: write the resulting files as a single tarball (.tar).
+
+If no type is specified, the value defaults to local. Use a hyphen as non-comma format (`-`) to write the output tarball to standard output (`STDOUT`).
```
This
- doesn’t document `dest=-` AFAICS, although the implementation now supports that.
- Says “non-comma format”, which is a short-hand I started using for the purposes of this review, but that is unclear without extra content.


Maybe something vaguely like
> ```
> - **dest**: … absolute or relative path; or, as a special case, `-` means the standard output.
> …
> Alternatively, instead of a comma-separated sequence, the value of **--output** can be just a destination (in the `**dest** format) (e.g. `--output some-path`, `--output -`)
> ```

mtrmac replied:
The “alternatively” part should also document the implied `type` value…

… wait, `--output foo` means `type=local`, but `--output -` means `type=tar`? That needs to be explicitly documented as well.

mtrmac commented on pkg/parse/parse.go:
```diff
@@ -510,6 +510,70 @@ func AuthConfig(creds string) (*types.DockerAuthConfig, error) {
+		return define.BuildOutputOption{}, fmt.Errorf("invalid build output options %q, accepted keys are type and dest must be present", buildOutput)
+	}
+
+	if !isDir && path == "-" {
```
Should we reject `type=dir,dest=-`?

mtrmac commented on pkg/parse/parse.go:
```diff
@@ -510,6 +510,70 @@ func AuthConfig(creds string) (*types.DockerAuthConfig, error) {
+// GetBuildOutput is responsible for parsing custom build output argument i.e `build --output` flag.
+// Takes `buildOutput` as string and returns BuildOutputOptions
+func GetBuildOutput(buildOutput string) (define.BuildOutputOption, error) {
+	isDir := true
+	isStdout := false
+	typeSelected := false
+	pathSelected := false
+	path := ""
```
(Absolutely non-blocking: All of this can now be moved just before the `tokens :=` initializer, requiring a tiny bit less context to keep in mind when reading the code.)

mtrmac commented on pkg/parse/parse.go:
```diff
@@ -510,6 +510,70 @@ func AuthConfig(creds string) (*types.DockerAuthConfig, error) {
 }
 
+// GetBuildOutput is responsible for parsing custom build output argument i.e `build --output` flag.
+// Takes `buildOutput` as string and returns BuildOutputOptions
```
```suggestion
// Takes `buildOutput` as string and returns BuildOutputOption
```

flouthoc commented on internal/util/util.go:
```diff
@@ -22,3 +30,35 @@ func LookupImage(ctx *types.SystemContext, store storage.Store, image string) (*
+		if err != nil {
+			return err
+		}
+		opts.Path = filepath.Join(cwd, opts.Path)
```
switched to abs(

mtrmac commented on docs/buildah-build.1.md:
```diff
@@ -411,6 +411,25 @@ By default, Buildah manages _/etc/hosts_, adding the container's own IP address.
+- **tar**: write the resulting files as a single tarball (.tar).
+
+If no type is specified, the value defaults to **local**.
+Alternatively, instead of a comma-separated sequence, the value of **--output** can be just a destination (in the `**dest** format) (e.g. `--output some-path`, `--output -`) where `--output some-path` is equivalent to **type=local** and `--output -` is equivalent to **type=tar**.
```
One last nit: `s/is equivalent to/is treated as if/` or something like that. `some-path` is not “equivalent” to a type value.

Or commit to this approach, and write “`--output some-path` is equivalent to `type=local,dest=some-path`”.

mtrmac commented on pkg/parse/parse.go:
```diff
@@ -510,6 +510,74 @@ func AuthConfig(creds string) (*types.DockerAuthConfig, error) {
+		return define.BuildOutputOption{}, fmt.Errorf("invalid build output option %q, accepted keys are type and dest must be present", buildOutput)
+	}
+
+	if !isDir && path == "-" {
+		return define.BuildOutputOption{Path: "",
+			IsDir:    false,
+			IsStdout: true}, nil
+	}
+
+	if isDir && path == "-" {
```
This can be simplified to
```go
if path == "-" {
   if isDir { fail }
   return $accepted
}
```
where the $accepted value is known to be correct by construction, and we don’t have to worry about the two conditions getting out of sync.

mtrmac reviewed (commented):
Thanks!

The local implementation LGTM now; some resolution of the ownership/set-UID behavior, by someone who understands it (i.e. not me) is still a blocker for me.

flouthoc commented:
@nalind Resolved all issues but this one is still left https://github.com/containers/buildah/pull/3823#discussion_r853821897

nalind commented on imagebuildah/stage_executor.go:
```diff
@@ -1556,6 +1567,25 @@ func (s *StageExecutor) commit(ctx context.Context, createdBy string, emptyLayer
+		if err != nil {
+			return "", nil, errors.Wrapf(err, "failed to export build output")
+		}
+		if errChan != nil {
```
This is dead code now, isn't it?

flouthoc commented on image.go:
```diff
@@ -813,3 +813,12 @@ func (b *Builder) makeImageRef(options CommitOptions) (types.ImageReference, err
+func (b *Builder) ExtractRootfs(options CommitOptions) (io.ReadCloser, chan error, error) {
+	src, err := b.makeContainerImageRef(options)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "error computing layer digests and building metadata for container %q", b.ContainerID)
```
fixed

nalind commented on image.go:
```diff
@@ -150,7 +150,10 @@ func computeLayerMIMEType(what string, layerCompression archive.Compression) (om
+// Takes copierStripOpts as argument which allows caller to configure
+// preserve nature of setuid,setgid,sticky and extended attributes
+// on extracted rootfs.
+func (i *containerImageRef) extractRootfs(copierStripOpts copier.GetOptions) (io.ReadCloser, chan error, error) {
```
Go ahead and define a `define.ExtractRootfsOptions` struct that only includes fields whose values are used.

nalind commented on imagebuildah/stage_executor.go:
```diff
@@ -1556,6 +1568,33 @@ func (s *StageExecutor) commit(ctx context.Context, createdBy string, emptyLayer
+			// Strip setuid,setgid,sticky and extended attributes
+			copierStripOpts.StripSetuidBit = true
+			copierStripOpts.StripSetgidBit = true
+			copierStripOpts.StripStickyBit = true
```
Looking at the top-level `tmp` directory produced by buildkit, I think `StripStickyBit` needs to be left `false` always, and as a result not be part of a `ExtractRootfsOptions` struct.

nalind commented on image.go:
```diff
@@ -43,6 +43,15 @@ const (
+type ExtractRootfsOptions struct {
+	StripSetuidBit bool // strip the setuid bit off of items being copied. no effect on archives being extracted
+	StripSetgidBit bool // strip the setgid bit off of items being copied. no effect on archives being extracted
+	StripXattrs    bool // don't record extended attributes of items being copied. no effect on archives being extracted
```
These comments need to be updated.

nalind commented on internal/util/util.go:
```diff
@@ -22,3 +31,45 @@ func LookupImage(ctx *types.SystemContext, store storage.Store, image string) (*
+
+		err = os.MkdirAll(opts.Path, 0700)
+		if err != nil {
+			return err
```
`errors.Wrapf()` this, so that if we fail here, the caller can get a stack trace that points to this line.

nalind commented:
Two nits, otherwise LGTM.

flouthoc commented:
@rhatdan @nalind needs /lgtm

mtrmac commented on imagebuildah/stage_executor.go:
```diff
@@ -1556,6 +1568,32 @@ func (s *StageExecutor) commit(ctx context.Context, createdBy string, emptyLayer
+	if s.executor.buildOutput != "" {
+		extractRootfsOpts := buildah.ExtractRootfsOptions{}
+		if unshare.IsRootless() {
+			// Strip setuid,setgid,sticky and extended attributes
```
Non-blocking? It’s kinda obvious that that’s what the three lines do, to me that doesn’t need a comment.

But I’d _love_ to see a comment recording the outcome of the discussions. What's the rationale for stripping these bits for rootless use, and not as root? (IOW, if someone came 5 years later thinking that this is surprising, what would be necessary to know?)

[OTOH if this is a hard-won consensus but asking for details would break that consensus :), by all means let’s leave it as it is.]

mtrmac commented on internal/util/util.go:
```diff
@@ -22,3 +31,45 @@ func LookupImage(ctx *types.SystemContext, store storage.Store, image string) (*
+		}
+	}
+	if opts.IsDir {
+		// preserve ownership when invoked as root
```
Same documentation point here.

(And if the two are parts of a single idea, as it seems to be the case, it would be nice to at least have comments that link the two. Or perhaps an explicit code link — have a `preserveOwnership bool` parameter to this function, and have the caller set it along with the three set-UID/set-GUID/xattr choices, in a single `if` block.)

flouthoc replied:
Added a reason and link to the thread.

mtrmac reviewed (commented):
Code LGTM, just stirring up trouble… Feel free to ignore me.

rhatdan commented:
/lgtm
/hold
@flouthoc did you open a bug/question on why buildkit does this change between rootless and rootful?
